### ISSUE TITLE
Switch secp256k1 dependency

### ISSUE THREAD

**Issue by ofek** [JoinMarket-Org/joinmarket#728]
Consider switching from [https://github.com/ludbb/secp256k1-py](https://github.com/ludbb/secp256k1-py) to [https://github.com/ofek/coincurve](https://github.com/ofek/coincurve) as the former is no longer maintained.

New features include:

- Support for Windows (Python 3 only though, unlike Linux/macOS)
- Linux, macOS, and Windows all have binary packages for both 64 and 32-bit architectures
- A fix to remove CFFI warnings
- A global context is used by default, drastically increasing performance
- Each release uses newest version of [libsecp256k1](https://github.com/bitcoin-core/secp256k1)
- Implements a fix for [https://bugs.python.org/issue28150](https://bugs.python.org/issue28150)  to support Python 3.6+ on macOS
- Linux & macOS wheels now use GMP
- Endomorphism optimization is now enabled

When you do switch, **use coincurve version 2.1.0 specifically!** Anything after that will have a modified (cleaner :) API and only support Python 3.

**Comment by AdamISZ** [comment]
@ofek thanks.

If someone wants to pursue this: would suggest making a branch to make this as an optional alternative for building initially, then people can test it out easily.

**Comment by mb300sd** [comment]
+1 

The old binding no longer runs with the latest build if libsecp256k1.

I tried and failed at converting, api changes too much for my minimal understanding of EC.

**Comment by AdamISZ** [comment]
@mb300sd thanks for the heads up. I will take a look today.

Is it libsecp256k1 rather than just secp256k1-py that has changed? (I guess so from the way you worded it). If so yeah it'll take some work and be quite hard to deal with since they don't issue releases. If it's secp256k1-py then we can just fix a particular version to install, for now.

**Comment by ofek** [comment]
@AdamISZ Yes libsecp256k1 has received many improvements since the last release of secp256k1-py, 6 performance enhancements just in the past few months https://github.com/ofek/coincurve#500

Have you taken a look at https://github.com/ofek/coincurve yet?

**Comment by mb300sd** [comment]
libsecp256k1 removed some functions related to schnorr signatures a while ago, so it fails with a cffi not found error when using the latest versions.

It looks simple enough to convert to coincurve, (for someone that understands the EC math), I got about half way though in an hour and couldn't tell what needed to be PrivateKey.multiply vs PublicKey.multiply anymore and gave up.

**Comment by AdamISZ** [comment]
@mb300sd 

Yes the schnorr stuff was removed but that shouldn't cause a problem iirc.

I did two experiments to verify what I believed to be the case. In both, I took a *completely fresh* VM install of Ubuntu 1404:

1. Build libsecp256k1 from source, then install secp256k1-py with the ridiculous syntax `INCLUDE_DIR=.. LIB_DIR=.. pip install --no-binary secp256k1 secp256k1`. (This was fiddly though e.g. needs LD_LIBRARY_PATH to link)). Verify with ldd/readelf that you are actually linking to the underlying. Then install joinmarket-clientserver (no different from this repo in this context)

2. Don't do the above, so that just `pip install secp256k1` has to source libsecp256k1 itself, before installing jmcs.

In neither case did I encounter problems with the code, once installed. I was able to run wallet-tool and get output.

Now this doesn't necessarily prove anything, but: if necessary, could you open a separate issue detailing exactly your situation: OS, what steps you followed, and especially: what error messages you got. And, sorry if I do not necessarily going to respond fast; I already spent 2-3 hours to this when I wanted to make more progress on the joinmarket-Qt code for the new segwit version. It's just a shame no one else is helping with that, else I'd feel less time pressure.

As to the topic of *this* issue, I'd still repeat what I said at the start, if someone wants to fork and make a branch using coincurve, they can report back on it, that'd be great. I get it that since the secp256k1-py is inactive, there can well be issues and we may have to change, but it will not be a minor piece of work.

**Comment by mb300sd** [comment]
The error I've been getting is 

```
  File "/usr/local/lib/python2.7/dist-packages/secp256k1/__init__.py", line 5, in <module>
    from ._libsecp256k1 import ffi, lib
ImportError: /usr/local/lib/python2.7/dist-packages/secp256k1/_libsecp256k1.so: undefined symbol: secp256k1_schnorr_verify
```

I'm running on Debian Jessie, with libsecp256k1 installed system wide from github. 04c8ef3 is the lastest version that doesn't cause the error. I have it working on my system, just posting for anyone else that happens to google the issue.
